libs-scala/resources/src/main/scala/com/digitalasset/resources/Resource.scala

```diff
@@ -58,33 +88,40 @@ trait Resource[+A] {
         release _,
       )
       .flatten
-
 }
 
 object Resource {
   import scala.language.higherKinds
 
+  /**
+    * Nests release operation for a [[Resource]]'s future.
+    */
   private def nest[T](future: Future[T])(
       releaseResource: T => Future[Unit],
       releaseSubResources: () => Future[Unit],
   )(implicit executionContext: ExecutionContext): Resource[T] =
     new Resource[T] {
-      private val released: AtomicBoolean = new AtomicBoolean(false)
-      private val releasePromise: Promise[Unit] = Promise()
 
       final lazy val asFuture: Future[T] = future.transformWith {
         case Success(value) => Future.successful(value)
-        case Failure(throwable) => release().flatMap(_ => Future.failed(throwable))
+        case Failure(throwable) =>
+          release().flatMap(_ => Future.failed(throwable)) // Release all if failing
       }
 
+      private val released: AtomicBoolean = new AtomicBoolean(false) // Short-circuits to a promise
+      private val releasePromise: Promise[Unit] = Promise() // Will be the release return handle
+
+      // Explicit release must call both main and nested releases
```

Comment:
I think you can delete this comment and make it part of the Scaladoc of the trait.